README.md
# ProgressPrinter [![Gem Version](https://badge.fury.io/rb/progress_printer.svg)](https://badge.fury.io/rb/progress_printer)

> Logs the progress of an operation, with estimated completion time.

## Installation

When using [Bundler](https://bundler.io), add this to your project's `Gemfile`:

```ruby
gem 'progress_printer'
```

Otherwise, install it with the `gem` command:


```shell
$ gem install progress_printer
```

### Already in a console?

ProgressPrinter is in a single file with no dependencies, so you're able to copy/paste the entire file into an `irb` session, and then immediately use `ProgressPrinter.new` (see below). The raw source to copy is available here:

[https://raw.githubusercontent.com/justincampbell/progress_printer/master/lib/progress_printer.rb](https://raw.githubusercontent.com/justincampbell/progress_printer/master/lib/progress_printer.rb)

## Usage

### Basic Usage

A `ProgressPrinter` must be created, started, and finished. Use `#increment` within your operation to increment the progress.

```rb
require 'progress_printer'

printer = ProgressPrinter.new(name: "Counting", total: 250)
printer.start
250.times { sleep 0.05; printer.increment }
printer.finish
```

Output:

```
Counting:   0/250   0% calculating...
Counting: 100/250  40% ~8s
Counting: 200/250  80% ~2s
Counting: 250/250 100% 14s total
```

You can also achieve the same results by using `.wrap` or `#wrap`:

```rb
ProgressPrinter.wrap(name: "Counting", total: 250) do |progress|
  250.times { sleep 0.05; progress.increment }
end
```

```rb
printer = ProgressPrinter.new(name: "Counting", total: 250)
printer.wrap do |progress|
  250.times { sleep 0.05; progress.increment }
end
```

### Arguments

* `total` - The total number of iterations expected. If this is omitted, estimated completion time will not be shown.
* `name` - A string to display next to each printed line. This helps identify the current operation, or the specific progress printer if using multiple.
* `every` (Default: `100`) - How many iterations should pass in between printing a line.
* `out` (Default: `$stdout`) - An object responding to `#puts` for printing the progress to.

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/justincampbell/progress_printer. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.


CHANGELOG.md
## 0.2.1

* Fix ProgressPrinter.wrap error on newer Ruby versions [[50d763f](https://github.com/justincampbell/progress_printer/commit/50d763f)]

## 0.2.0

* Print the total runtime on #finish [[PR 6](https://github.com/justincampbell/progress_printer/pull/6/files)]
* Added .wrap/#wrap methods which take blocks [[PR 4](https://github.com/justincampbell/progress_printer/pull/4/files)]

## 0.1.0

* Initial release



spec/spec_helper.rb
require "bundler/setup"
require "progress_printer"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/progress_printer_spec.rb
require 'spec_helper'
require 'progress_printer'

RSpec.describe ProgressPrinter do
  subject(:printer) {
    described_class.new(
      name: name,
      every: every,
      total: total,
      out: output
    )
  }

  let(:output) { StringIO.new }
  let(:total) { nil }
  let(:name) { nil }
  let(:every) { described_class::DEFAULT_EVERY }

  before do
    allow_any_instance_of(ProgressPrinter)
      .to receive(:testing?)
      .and_return(false)
  end

  context "example with all options" do
    let(:total) { 250 }
    let(:name) { "Counting" }
    let(:every) { 100 }

    it "prints the progress of a loop" do
      printer.start
      250.times { printer.increment }
      printer.finish
      expect(output.string).to eq(
        <<-OUTPUT.gsub(/^\s*/, '')
          Counting:   0/250   0% calculating...
          Counting: 100/250  40% ~0s
          Counting: 200/250  80% ~0s
          Counting: 250/250 100% ~0s
        OUTPUT
      )
    end
  end

  describe ".format_duration" do
    {
      0 => "0s",
      1 => "1s",
      60 => "1m",
      90 => "1m30s",
      3600 => "1h",
      9030 => "2h30m30s",
      604800 => "7d",
      1234567890 => "14288d23h31m30s"
    }.each do |seconds, expected|
      it "converts #{seconds} seconds to #{expected}" do
        expect(described_class.format_duration(seconds)).to eq(expected)
      end
    end
  end

  describe ".wrap" do
    it "creates a new instance with all args and wraps the given block" do
      described_class.wrap name: "Count", every: 1, total: 1, out: output do |progress|
        progress.increment
      end

      expect(output.string).to eq(
        <<-OUTPUT.gsub(/^\s*/, '')
          Count: 0/1   0% calculating...
          Count: 1/1 100% ~0s
          Count: 1/1 100% ~0s
        OUTPUT
      )
    end
  end

  describe "#wrap" do
    it "starts and finishes around a given block" do
      expect(printer).to receive(:start).and_call_original
      expect(printer).to receive(:finish).and_call_original

      printer.wrap do |printer|
        printer.increment
      end

      expect(output.string).to eq("0\n1\n")
    end

    it "returns the value of the block" do
      result = printer.wrap { :abc }

      expect(result).to eq(:abc)
    end
  end

  describe "#start" do
    it "prints 0" do
      printer.start
      expect(output.string).to eq("0\n")
    end

    context "with a name" do
      let(:name) { "Hard Work" }

      it "prefixes the name" do
        printer.start
        expect(output.string).to eq("Hard Work: 0\n")
      end
    end

    context "with a total" do
      let(:total) { 5 }

      it "prints 0 out of total" do
        printer.start
        expect(output.string).to eq("0/5   0% calculating...\n")
      end

      context "with a longer total" do
        let(:total) { 12345 }

        it "left-pads the progress" do
          printer.start
          expect(output.string).to eq("    0/12345   0% calculating...\n")
        end
      end
    end
  end

  describe "#increment" do
    it "does not print before reaching the every number" do
      99.times { printer.increment }
      expect(output.string).to eq("")
    end

    it "prints when reaching the every number" do
      100.times { printer.increment }
      expect(output.string).to eq("100\n")
    end

    context "with a count" do
      it "increments by that count" do
        printer.increment(123)
        expect(printer.current).to eq(123)
      end
    end
  end

  describe "#finish" do
    it "prints the current number" do
      123.times { printer.increment }
      output.reopen
      printer.finish
      expect(output.string).to eq("123\n")
    end

    context "with a total" do
      let(:total) { 123 }

      it "prints the total" do
        printer.finish
        expect(output.string).to eq("123/123   0% calculating...\n")
      end
    end

    context "when the total was already printed" do
      before do
        100.times { printer.increment }
        expect(output.string).to eq("100\n")
        output.reopen
      end

      it "does not print the total again" do
        printer.finish
        expect(output.string).to eq("")
      end
    end
  end

  describe "#percent_complete" do
    subject(:percent_complete) { printer.percent_complete }

    it { is_expected.to eq(nil) }

    context "with a total" do
      let(:total) { 20 }

      it { is_expected.to eq(0.0) }

      context "when a quarter done" do
        before { (total / 4).times { printer.increment } }

        it { is_expected.to eq(0.25) }
      end

      context "when over incremented" do
        before { (total + 1).times { printer.increment } }

        it { is_expected.to eq(1.0) }
      end
    end

    context "with a zero total" do
      let(:total) { 0 }

      it { is_expected.to eq(1.0) }
    end
  end

  describe "#percent_complete_string" do
    subject(:percent_complete_string) { printer.percent_complete_string }

    it { is_expected.to eq(nil) }

    context "with a total" do
      let(:total) { 20 }

      it { is_expected.to eq("0%") }

      context "when a quarter done" do
        before { (total / 4).times { printer.increment } }

        it { is_expected.to eq("25%") }
      end

      context "when over incremented" do
        before { (total + 1).times { printer.increment } }

        it { is_expected.to eq("100%") }
      end
    end

    context "with a zero total" do
      let(:total) { 0 }

      it { is_expected.to eq("100%") }
    end
  end

  describe "#percent_remaining" do
    subject(:percent_remaining) { printer.percent_remaining }

    it { is_expected.to eq(nil) }

    context "with a total" do
      let(:total) { 20 }

      it { is_expected.to eq(1.0) }

      context "when a quarter done" do
        before { (total / 4).times { printer.increment } }

        it { is_expected.to eq(0.75) }
      end

      context "when over incremented" do
        before { (total + 1).times { printer.increment } }

        it { is_expected.to eq(0.0) }
      end
    end

    context "with a zero total" do
      let(:total) { 0 }

      it { is_expected.to eq(0.0) }
    end
  end

  describe "#estimated_time_remaining" do
    subject(:estimated_time_remaining) {
      printer.estimated_time_remaining(current_time)
    }

    let(:start_time) { Time.new(2016, 1, 2, 3, 4, 5, 6) }
    let(:current_time) { start_time }

    before { printer.start_time = start_time }

    it { is_expected.to eq(nil) }

    context "with a total" do
      let(:total) { 20 }

      it { is_expected.to eq("calculating...") }

      context "when 10 percent done" do
        before { (total / 10).times { printer.increment } }

        it { is_expected.to eq("~0s") }

        context "when 1 second has passed" do
          let(:current_time) { start_time + 1 }

          it { is_expected.to eq("~9s") }
        end

        context "when 10 seconds have passed" do
          let(:current_time) { start_time + 10 }

          it { is_expected.to eq("~1m30s") }
        end
      end

      context "when 50 percent done" do
        before { (total / 2).times { printer.increment } }

        it { is_expected.to eq("~0s") }

        context "when 1 second has passed" do
          let(:current_time) { start_time + 1 }

          it { is_expected.to eq("~1s") }
        end

        context "when 10 seconds have passed" do
          let(:current_time) { start_time + 10 }

          it { is_expected.to eq("~10s") }
        end
      end

      context "when done" do
        before { total.times { printer.increment } }

        it { is_expected.to eq("~0s") }

        context "when 1 second has passed" do
          let(:current_time) { start_time + 1 }

          it { is_expected.to eq("~0s") }
        end

        context "when 10 seconds have passed" do
          let(:current_time) { start_time + 10 }

          it { is_expected.to eq("~0s") }
        end
      end
    end
  end

  describe "#seconds_remaining" do
    subject(:seconds_remaining) { printer.seconds_remaining(current_time) }

    let(:start_time) { Time.new(2016, 1, 2, 3, 4, 5, 6) }
    let(:current_time) { start_time }

    before { printer.start_time = start_time }

    it { is_expected.to eq(nil) }

    context "with a total" do
      let(:total) { 20 }

      it { is_expected.to eq(nil) }

      context "when 10 percent done" do
        before { (total / 10).times { printer.increment } }

        it { is_expected.to eq(0.0) }

        context "when 1 second has passed" do
          let(:current_time) { start_time + 1 }

          it { is_expected.to eq(9.0) }
        end

        context "when 10 seconds have passed" do
          let(:current_time) { start_time + 10 }

          it { is_expected.to eq(90.0) }
        end
      end

      context "when 50 percent done" do
        before { (total / 2).times { printer.increment } }

        it { is_expected.to eq(0.0) }

        context "when 1 second has passed" do
          let(:current_time) { start_time + 1 }
          it { is_expected.to eq(1.0) }
        end

        context "when 10 seconds have passed" do
          let(:current_time) { start_time + 10 }

          it { is_expected.to eq(10.0) }
        end
      end

      context "when done" do
        before { total.times { printer.increment } }

        it { is_expected.to eq(0.0) }

        context "when 1 second has passed" do
          let(:current_time) { start_time + 1 }
          it { is_expected.to eq(0.0) }
        end

        context "when 10 seconds have passed" do
          let(:current_time) { start_time + 10 }

          it { is_expected.to eq(0.0) }
        end
      end

      context "when over incremented" do
        before { (total + 1).times { printer.increment } }

        it { is_expected.to eq(0.0) }

        context "when 1 second has passed" do
          let(:current_time) { start_time + 1 }
          it { is_expected.to eq(0.0) }
        end

        context "when 10 seconds have passed" do
          let(:current_time) { start_time + 10 }

          it { is_expected.to eq(0.0) }
        end
      end
    end
  end
end


lib/progress_printer.rb
# A progress printer which simplifies logging the progress of loops. To use,
# create a new ProgressPrinter, and then increment inside of a loop.
#
# Example:
#
#   printer = ProgressPrinter.new(name: "Counting", total: 250)
#   printer.start
#   250.times { printer.increment }
#   printer.finish
#
# Output:
#
#   Counting:   0/250   0% calculating...
#   Counting: 100/250  40% ~1m30s
#   Counting: 200/250  80% ~30s
#   Counting: 250/250 100% 2m30s total
#
class ProgressPrinter
  DEFAULT_EVERY = 100

  class << self
    attr_accessor :silent

    def silence
      self.silent = true
    end

    def wrap(**args, &block)
      new(**args).wrap(&block)
    end
  end

  attr_reader :total, :name, :every, :out
  attr_accessor :start_time

  def initialize(total: nil, name: nil, every: DEFAULT_EVERY, out: $stdout)
    @total = total
    @name = name
    @every = every

    if self.class.silent
      @out = StringIO.new
    else
      @out = out
    end
  end

  def wrap
    start
    yield(self)
  ensure
    finish
  end

  def start
    self.start_time = Time.now
    print_progress 0
  end

  def increment(count = 1)
    n = nil
    count.times { n = enum.next }
    print_progress n if at_milestone?
  end

  def finish
    print_progress current, final: true
  end

  def percent_complete
    return unless total
    return 1.0 if current >= total
    current / total.to_f
  end

  def percent_complete_string
    return unless total
    "#{(percent_complete * 100).to_i}%"
  end

  def percent_remaining
    return unless total
    1.0 - percent_complete
  end

  def estimated_time_remaining(now = Time.now)
    return unless total
    return "calculating..." if current == 0

    "~" + self.class.format_duration(seconds_remaining(now))
  end

  def seconds_remaining(now)
    return unless total
    return if percent_remaining == 1.0
    return 0.0 if percent_remaining == 0.0

    time_passed(now) / percent_complete * (1.0 - percent_complete)
  end

  def time_passed(now = Time.now)
    return unless start_time
    now - start_time
  end

  def current
    enum.peek - 1
  end

  private

  def at_milestone?
    current % every == 0
  end

  def enum
    @enum ||= (1..Float::INFINITY).enum_for
  end

  def print_progress(n, final: false)
    buffer = StringIO.new

    buffer.print "#{name}: " if name

    if total
      buffer.print left_pad(n, total_length)
      buffer.print "/#{total} "
      buffer.print left_pad(percent_complete_string, 4)
    else
      buffer.print n
    end

    if final
      if time_passed
        buffer.print " #{self.class.format_duration(time_passed)} total"
      end
    else
      if total
        buffer.print " #{estimated_time_remaining}"
      end
    end

    out.puts buffer.string
  end

  def left_pad(string, length)
    pad_length = length - string.to_s.length
    return string if pad_length <= 0
    "#{' ' * pad_length}#{string}"
  end

  def total_length
    @total_length ||= total.to_s.length
  end

  def self.format_duration(seconds)
    seconds = seconds.to_i

    minutes = seconds / 60
    seconds = seconds % 60

    hours = minutes / 60
    minutes = minutes % 60

    days = hours / 24
    hours = hours % 24

    buffer = ""
    buffer.prepend "#{seconds.to_i}s" if seconds.nonzero?
    buffer.prepend "#{minutes.to_i}m" if minutes.nonzero?
    buffer.prepend "#{hours.to_i}h" if hours.nonzero?
    buffer.prepend "#{days.to_i}d" if days.nonzero?
    buffer = "0s" if buffer == ""
    buffer
  end
end
